cabal-install/Distribution/Client/ProjectConfig.hs

```diff
@@ -567,15 +578,50 @@ data BadPackageLocationMatch
   deriving Show
 
 renderBadPackageLocations :: BadPackageLocations -> String
-renderBadPackageLocations (BadPackageLocations bpls) =
-    unlines (map renderBadPackageLocation bpls)
+renderBadPackageLocations (BadPackageLocations prov bpls)
+      -- There is no provenance information,
+      -- render standard bad package error information.
+    | null provenance = renderErrors renderBadPackageLocation
+
+      -- The configuration is implicit, render bad package locations
+      -- using possibly specialized error messages.
+    | provenance == [Implicit] = renderErrors renderImplicitBadPackageLocation
+
+      -- The configuration was read from one or more explicit path(s),
+      -- list the locations and render the bad package error information.
+      -- The intent is to supersede this with the relevant location information
+      -- per package error.
+    | otherwise =
+           "When using configuration(s) from "
+        ++ intercalate ", " (mapMaybe getExplicit provenance)
+        ++ ", the following errors occurred:\n"
+        ++ renderErrors renderBadPackageLocation
+  where
+    renderErrors f = unlines (map f bpls)
+
+    provenance = Set.toList prov
+
+    getExplicit (Explicit path) = Just path
+    getExplicit Implicit        = Nothing
 
 --TODO: [nice to have] keep track of the config file (and src loc) packages
 -- were listed, to use in error messages
---TODO: [nice to have] keep track in the ProjectConfig and BadPackageLocations
--- of whether the project config was explicit or implicit so we can report a
--- better message, either pointing to the file or talking about the issues
--- related to having no project file and no package.
+
+-- | Render bad package location error information for the implicit
+-- @cabal.project@ configuration.
+--
+-- TODO: This is currently a placeholder with the intention to supply
+-- informative help text about the issues related to having no project
+-- file and no package.
+renderImplicitBadPackageLocation :: BadPackageLocation -> String
+renderImplicitBadPackageLocation bpl = case bpl of
+    BadLocGlobEmptyMatch pkglocstr ->
+        "No cabal.project or cabal file matching the glob '"
```

Comment:
Perhaps say "default glob" rather than "glob"? Because it doesn't actually exist anywhere.